src/compile/unit.ts

```diff
@@ -83,11 +83,18 @@ export class UnitModel extends Model {
 
   private _initScale(mark: Mark, encoding: Encoding, config: Config): Dict<Scale> {
     return UNIT_SCALE_CHANNELS.reduce(function(_scale, channel) {
-      if (vlEncoding.has(encoding, channel)) {
-        const scaleSpec = encoding[channel].scale || {};
-        const channelDef = encoding[channel];
+      if (vlEncoding.has(encoding, channel) ||
+          (channel === X && vlEncoding.has(encoding, X2)) ||
+          (channel === Y && vlEncoding.has(encoding, Y2))) {
+
+        let targetChannel = channel;
+        if (!vlEncoding.has(encoding, channel)) {
+          targetChannel = channel === X ? X2 : Y2;
+        }
+        const channelDef = encoding[targetChannel];
+        const scaleSpec = channelDef.scale || {};
 
-        const _scaleType = scaleType(scaleSpec, channelDef, channel, mark);
+        const _scaleType = scaleType(scaleSpec, channelDef, targetChannel, mark);
```

Comment:
Another part is that tricky part is here `channelDef` of `x` and `x2`  should have the same `type`, `bin`, `timeUnit`.   

Maybe we have to generate warning if they do not.